shotover/src/transforms/opensearch/mod.rs

```diff
@@ -0,0 +1,200 @@
+    address: String,
+    tls: Option<TlsConnector>,
+    connection: Option<Connection>,
+    connect_timeout: Duration,
+    codec_builder: OpenSearchCodecBuilder,
+}
+
+#[async_trait]
+impl Transform for OpenSearchSink {
+    async fn transform<'a>(&'a mut self, requests_wrapper: Wrapper<'a>) -> Result<Messages> {
+        // Return immediately if we have no messages.
+        // If we tried to send no messages we would block forever waiting for a reply that will never come.
+        if requests_wrapper.requests.is_empty() {
+            return Ok(requests_wrapper.requests);
+        }
+
+        if self.connection.is_none() {
+            trace!("creating outbound connection {:?}", self.address);
+
+            let generic_stream = if let Some(tls) = self.tls.as_mut() {
+                let tls_stream = tls
+                    .connect(self.connect_timeout, self.address.clone())
+                    .await?;
+                Box::pin(tls_stream) as Pin<Box<dyn AsyncStream + Send + Sync>>
+            } else {
+                let tcp_stream =
+                    tcp::tcp_stream(self.connect_timeout, self.address.clone()).await?;
+                Box::pin(tcp_stream) as Pin<Box<dyn AsyncStream + Send + Sync>>
+            };
+
+            let (decoder, encoder) = self.codec_builder.build();
+            let (stream_rx, stream_tx) = tokio::io::split(generic_stream);
+            let outbound_tx = FramedWrite::new(stream_tx, encoder);
+            let outbound_rx = FramedRead::new(stream_rx, decoder);
+            let (response_messages_tx, response_messages_rx) = mpsc::unbounded_channel();
+
+            tokio::spawn(
+                server_response_processing_task(outbound_rx, response_messages_tx)
+                    .in_current_span(),
+            );
+            self.connection = Some(Connection {
+                response_messages_rx,
+                outbound_tx,
+            })
+        }
+
+        let connection = self.connection.as_mut().unwrap();
+
+        let messages_len = requests_wrapper.requests.len();
+
+        let mut result = Vec::with_capacity(messages_len);
+        for message in requests_wrapper.requests {
+            connection
+                .outbound_tx
+                .send(vec![message])
+                .await
+                .map_err(|err| {
+                    anyhow!("Failed to send messages to OpenSearch destination: {err:?}")
+                })?;
+
+            let message = connection
+                .response_messages_rx
+                .recv()
+                .await
+                .ok_or_else(|| anyhow!("Failed to receive message because OpenSearchSink response processing task is dead"))?;
+            result.push(message);
+        }
+
+        Ok(result)
+    }
+}
+
+struct Connection {
+    outbound_tx: FramedWrite<WriteHalf<PinStream>, OpenSearchEncoder>,
+    response_messages_rx: mpsc::UnboundedReceiver<Message>,
+}
+
+async fn server_response_processing_task(
```

Comment:
why cant we just use `crate::transforms::util::cluster_connection_pool::spawn_read_write_tasks`?

Reply:
Because HTTP is a synchronous protocol, we want to send a message, wait for a response and then send the next message.

Reply:
spawn_read_write_tasks should work for both synchronous and asynchronous protocols.
we use it for kafka which is synchronous.

Reply:
Looking into this more, the current state of the sink is closer to `RedisSinkSingle` than the cluster implementation. I think keeping it simplest would be the best path forward for the time being.

Reply:
is this supposed to be a sink cluster or a single cluster?
We should name it either OpenSearchSinkSingle or OpenSearchSinkCluster to be consistent with our other sinks.

Reply:
I swapped this over to `spawn_read_write_tasks` and the tests still pass:
```diff
diff --git a/shotover/src/transforms/opensearch/mod.rs b/shotover/src/transforms/opensearch/mod.rs
index a91727a6..d7d8ca7d 100644
--- a/shotover/src/transforms/opensearch/mod.rs
+++ b/shotover/src/transforms/opensearch/mod.rs
@@ -1,13 +1,16 @@
-use crate::codec::{
-    opensearch::{OpenSearchCodecBuilder, OpenSearchDecoder, OpenSearchEncoder},
-    CodecBuilder, Direction,
-};
 use crate::message::Message;
 use crate::tcp;
 use crate::tls::AsyncStream;
 use crate::transforms::{
     Messages, Transform, TransformBuilder, TransformConfig, Transforms, Wrapper,
 };
+use crate::{
+    codec::{
+        opensearch::{OpenSearchCodecBuilder, OpenSearchDecoder, OpenSearchEncoder},
+        CodecBuilder, Direction,
+    },
+    transforms::util::cluster_connection_pool::spawn_read_write_tasks,
+};
 use anyhow::{anyhow, Result};
 use async_trait::async_trait;
 use futures::{FutureExt, SinkExt, StreamExt};
@@ -15,10 +18,13 @@ use serde::Deserialize;
 use std::pin::Pin;
 use std::time::Duration;
 use tokio::io::{ReadHalf, WriteHalf};
-use tokio::sync::mpsc;
+use tokio::sync::{mpsc, oneshot};
 use tokio_util::codec::{FramedRead, FramedWrite};
 use tracing::{trace, Instrument};
 
+use super::util::cluster_connection_pool::Connection;
+use super::util::Request;
+
 type PinStream = Pin<Box<dyn AsyncStream + Send + Sync>>;
 
 #[derive(Deserialize, Debug)]
@@ -96,22 +102,9 @@ impl Transform for OpenSearchSink {
             trace!("creating outbound connection {:?}", self.address);
 
             let tcp_stream = tcp::tcp_stream(self.connect_timeout, self.address.clone()).await?;
-            let stream = Box::pin(tcp_stream) as Pin<Box<dyn AsyncStream + Send + Sync>>;
-
-            let (decoder, encoder) = self.codec_builder.build();
-            let (stream_rx, stream_tx) = tokio::io::split(stream);
-            let outbound_tx = FramedWrite::new(stream_tx, encoder);
-            let outbound_rx = FramedRead::new(stream_rx, decoder);
-            let (response_messages_tx, response_messages_rx) = mpsc::unbounded_channel();
-
-            tokio::spawn(
-                server_response_processing_task(outbound_rx, response_messages_tx)
-                    .in_current_span(),
-            );
-            self.connection = Some(Connection {
-                response_messages_rx,
-                outbound_tx,
-            })
+            let (rx, tx) = tcp_stream.into_split();
+
+            self.connection = Some(spawn_read_write_tasks(&self.codec_builder, rx, tx));
         }
 
         let connection = self.connection.as_mut().unwrap();
@@ -120,61 +113,18 @@ impl Transform for OpenSearchSink {
 
         let mut result = Vec::with_capacity(messages_len);
         for message in requests_wrapper.requests {
+            let (tx, rx) = oneshot::channel();
             connection
-                .outbound_tx
-                .send(vec![message])
-                .await
-                .map_err(|err| {
-                    anyhow!("Failed to send messages to OpenSearch destination: {err:?}")
-                })?;
-
-            let message = connection
-                .response_messages_rx
-                .recv()
-                .await
-                .ok_or_else(|| anyhow!("Failed to receive message because OpenSearchSink response processing task is dead"))?;
+                .send(Request {
+                    message,
+                    return_chan: Some(tx),
+                })
+                .map_err(|_| anyhow!("Failed to send"))?;
+
+            let message = rx.await?.response?;
             result.push(message);
         }
 
         Ok(result)
     }
 }
-
-struct Connection {
-    outbound_tx: FramedWrite<WriteHalf<PinStream>, OpenSearchEncoder>,
-    response_messages_rx: mpsc::UnboundedReceiver<Message>,
-}
-
-async fn server_response_processing_task(
-    mut outbound_rx: FramedRead<ReadHalf<PinStream>, OpenSearchDecoder>,
-    response_messages_tx: mpsc::UnboundedSender<Message>,
-) {
-    loop {
-        tokio::select! {
-            responses = outbound_rx.next().fuse() => {
-                match responses {
-                    Some(Ok(messages)) => {
-                        for message in messages {
-                            if let Err(mpsc::error::SendError(_)) = response_messages_tx.send(message) {
-                                // RawSinkSingle dropped after a message was received from server,
-                                // request processor task shutting down
-                                return;
-                            }
-                        }
-                    }
-                    Some(Err(err)) => {
-                        tracing::error!("encountered error in OpenSearchSink stream: {err:?}");
-                        return;
-                    }
-                    None => {
-                        return;
-                    }
-                }
-            },
-            _ = response_messages_tx.closed() => {
-                // sink stream ended, shutdown the response processing task
-                return;
-            },
-        }
-    }
-}

```

Reply:
I'm happy to switch it over to avoid adding more lines of code, I just found the initial one easier to reason about.

Reply:
yep I agree that spawn_read_write_tasks has a very complicated implementation. :/
In the future we should investigate if we can simplify it, I'm thinking we might be able to change it to use a single task without a performance hit.

In the meantime we are better off using a known good implementation rather than reimplementing spawn_read_write_tasks.
Either we do end up simplifying it and then all sinks benefit from the simplification or we dont because the complexity was actually important.